### ISSUE TITLE
Missing lines in terminal scrollback

### ISSUE THREAD

**Issue by Zoybean** [nushell/nushell#9166]
### Describe the bug

I've noticed that whenever the prompt is printed, a number of lines are trimmed from the top of the screen.

unfortunately I haven't had the time to test thoroughly.

### How to reproduce

1. set a prompt taller than a single line
2. run a command with more lines of output than the terminal is tall (e.g. `seq 0 100`)
3. scroll up
4. observe that several lines of the command are missing

### Expected behavior

I expect that all lines of input and output are visible in the terminal history

### Screenshots

![image](https://github.com/nushell/nushell/assets/24929040/840bf908-644d-490c-b715-ca0b62764154)
![image](https://github.com/nushell/nushell/assets/24929040/c4247008-3cee-44fc-b4e2-9196742dd2a0)

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.75.1                                   |
| branch             | main                                     |
| commit_hash        | 1096e653b0b03728a7d5ba56fc3ad703395bcfbe |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.66.1 (90743e729 2023-01-10)      |
| rust_channel       | 1.66.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.66.1 (ad779e08b 2023-01-10)      |
| pkg_version        | 0.75.1                                   |
| build_time         | 2023-02-12 19:31:10 +08:00               |
| build_rust_channel | release                                  |
| features           | database, default, trash, which, zip     |
| installed_plugins  |                                          |

### Additional context

Possibly related to nushell/reedline#620

**Comment by amtoine** [comment]
hey @Zoybean 
could you try the same steps with the newest version, to see if it's fixed? :relieved: 

also i cannot reproduce this with my `alacritty` terminal emulator... that might be an issue with your terminal? :open_mouth: 
which one is it?

**Comment by lodezo** [comment]
A comparable issue is recently created in the Microsoft (Windows) Terminal project:
https://github.com/microsoft/terminal/issues/15330

It seems totally unrelated to nushell.

**Comment by Zoybean** [comment]
Thanks for the quick response; I'm indeed on windows terminal, and trying the same basic test on `nu` on `cmd` doesn't reproduce the issue. Closing

**Comment by Zoybean** [comment]
While researching more to comment on the Windows Terminal issue linked above, I updated nushell (on WSL and on the host) to use the version at the latest `master` (8d8304cf9108bcd6cb9b40a90a4bc92364000e92), and found that I can now reproduce this in `nu` on any terminal I have available locally. I don't know why, but I guess the old version I had on the host was not exhibiting the issue. Unfortunately, I do not recall what version it was previously.

- terminal emulators: windows terminal (using WSL), cmd, conemu
- shells: nu, bash, powershell

I was only able to reproduce in nushell, in any of the above terminals.

Here's a script you can run to test in `nu`:
```nu
let-env N = 5
let-env PROMPT_COMMAND = { seq 0 $env.N | str join "\n" }
seq 0 ((term size | get rows | into int) - 2)
```
After running, the first line of scrollback is always the seq output at 0, but the line after is N+1. This is the case regardless of the terminal size and the value of `$env.N`

e.g. for N=5, rows=14. line 6 is the top of the terminal window before scrolling:
```
5> seq 0 ((term size | get rows | into int) - 2)
╭────┬────╮
│  0 │  0 │
│  6 │  6 │
│  7 │  7 │
│  8 │  8 │
│  9 │  9 │
│ 10 │ 10 │
│ 11 │ 11 │
│ 12 │ 12 │
╰────┴────╯
0
2
3
4
5>
```

---

To check that it doesn't occur in bash or powershell, here's scripts for those:
`bash`:
```bash
export N=5
export PS1=$'$(seq 0 $N)> '
echo --; seq 0 $(($(stty size | cut -d ' ' -f1)-2)); echo --
```
`powershell` 6.2 or later:
```powershell
$env:N = 5
function Prompt { $(0..$env:N | Join-String -Separator "`r`n") + "> "}
echo "--"; 0..$Host.UI.RawUI.WindowSize.Height; echo "--"
```

**Comment by fdncred** [comment]
I still can't reproduce this using WezTerm on MacOS.
The way the terminal looks when finished with seq.
![image](https://github.com/nushell/nushell/assets/343840/b86a7c91-2d0b-43e9-93fa-c3b91ea39906)
And scrolled up
![image](https://github.com/nushell/nushell/assets/343840/5d9a397d-8e8d-408d-a22e-651ef31ac6fe)

**Comment by Zoybean** [comment]
yeah you've definitely tested right as far as my method goes, and your set-up definitely isn't exhibiting the issue. I wouldn't be surprised if it was somehow windows specific.

**Comment by lodezo** [comment]
I was able to further pinpoint this and found a workaround: When I disable Starship, by removing it from config.nu and restart the shell the problem dissapears. I tested this quite a few times. 

I guess you use Starship, based on the screenshot with fancy prompt? Can you please try to disable Starship or run config reset to be sure?

Some further info:

I was able to reproduce this behaviour with:
- different versions of Windows Terminal (1.15 (old), 1.16 (current), 1.17 (preview).
- different terminal sizes, fonts and appearance settings
- Windows native and WSL linux
- Nu started on it's own or from Powershell
- simple commands like `2..80` and `help commands`
- table and text output like `2..80 | to text`

Than I found it can also be reproduced in the Alacritty and VSCode integrated terminals. The missing lines are always the last ones that scrolled away from the screen.

I furthermore tried some different Starship themes and settings, without results.

**Comment by Zoybean** [comment]
Just managed to confirm that it is not Windows/WSL-linux specific; I can reproduce it on native Linux Mint in gnome-terminal, built locally to the same version I mentioned above (8d8304cf9108bcd6cb9b40a90a4bc92364000e92).
I do indeed use starship, but (still on linux) running `config reset` and opening a new shell instance do not seem to solve the issue.

**Comment by lodezo** [comment]
I just installed a fresh copy of Manjaro linux in Virtualbox. I installed nushell and Starship via the pacman package manager. I used the built-in Konsole terminal and Alacritty and was not able to reproduce the issue with or without Starship.

Back in Windows Terminal I was able to reproduce the issue directly after sourcing the Starship init.nu file. 

I activated the Windows Terminal debug mode in which also non visible output can be seen. I found that after activating Starship the output changes. The `2..80` output table is printed fully and is printed again partly. 

␛[37m│␣␛[32m␛[1m77␛[m␣␛[37m│␣79␣│␛[K␛[m␍␊
␛[37m│␣␛[32m␛[1m78␛[m␣␛[37m│␣80␣│␛[K␛[m␍␊
␛[37m╰────┴────╯␛[m␍␊
␛[5␣q␛[?25l␛[37m␛[H│␣␛[32m␛[1m47␛[m␣␛[37m│␣49␣│␛[K␍␊
│␣␛[32m␛[1m48␛[m␣␛[37m│␣50␣│␛[K␍␊
│␣␛[32m␛[1m49␛[m␣␛[37m│␣51␣│␛[K␍␊

**Comment by fdncred** [comment]
@lodezo how do you enable this functionality in WT?

**Comment by lodezo** [comment]
Instructions for enabling debug mode can be found [here](https://github.com/microsoft/terminal/wiki/Enabling-the-debug-tap).

Normally this is the location of the Windows Terminal settings file:
~\AppData\Local\Packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings.json

- Add "debugFeatures": true
- Restart the terminal
- Press both alt keys when starting a new Nushell tab

Focus on the debug screen, rightclick on the tab and choose export Text to create a text file with the debug results.

**Comment by lhecker** [comment]
I work on Windows Terminal and am slowly starting to work on this issue (lined above). A colleague hinted that we had a similar issue filed on us before: https://github.com/PowerShell/PSReadLine/issues/724

It was resolved in PSReadLine itself. Now I just noticed that nushell does the exact same thing and notes it in a comment as well: https://github.com/nushell/reedline/blob/7f54706cab9c64139e382936303a484c9c2726d3/src/painting/painter.rs#L131-L132
```rs
    /// Note. The `ScrollUp` operation in `crossterm` deletes lines from the top of
    /// the screen.
    pub(crate) fn repaint_buffer(
```

The call to `crossterm`'s `ScrollUp` boils down to `format!(\x1b[{}S, extra)`. I wonder if this is related and optimally would need to be resolved on nushell's side as well? 🤔 I'm honestly not entirely sure yet what's happening...

**Comment by fdncred** [comment]
@lhecker Thanks for your input! I'm excited to have someone else looking at this with "new eyes".

One of the problems that we've had is reproducing it on demand. We've definitely seen this issue in Windows Terminal and VSCode Terminal. I suspect we have a few more issues filed in our repo about it as well. I use a variety of terminals on macos, windows, and to a lesser degree, linux and I just haven't got very far with it. So, I'm pumped to learn how your research goes.

**Comment by j4james** [comment]
FYI, I looked into this a couple of years back, and the terminals at the time that matched the Windows Terminal behavior included Gnome Terminal (assumedly all VTE derivatives), Konsole, Hyper, and PowerTerm. There were also a whole lot of terminals that didn't support the `SU` sequence at all. Also worth mentioning that the original DEC hardware terminals used `SU` for something else entirely. The bottom line is that sequence is best avoided.

**Comment by j4james** [comment]
@lodezo I believe VSCode and Alacritty both use conpty on Windows, and that's where the scrolling is handled, so I'd expect them to share the same behavior as Windows Terminal (i.e. the output is expected to be truncated).

Alacritty should be fine on Linux, but I suspect VSCode will still truncate.

**Comment by lodezo** [comment]
@j4james  I just tried them all, you're right.

So this is the outcome, just with the reedline basic example:
| Terminal         | Windows | Linux |
| ---------------- | ------- | ----- |
| Alacritty        | Truncate     | No    |
| VSCode           | Truncate     | Truncate   |
| Konsole          | N/A     | No    |
| Windows Terminal | Truncate     | N/A   |

**Comment by j4james** [comment]
@lodezo I actually thought Konsole would have truncated too, but it looks like that's been fixed since I last tested (merge request [589](https://invent.kde.org/utilities/konsole/-/merge_requests/589)).

**Comment by NotLebedev** [comment]
Want to check in on this issue. Tested a little bit more and tried out some passable fixes.

First, I reproduced this issue on Alacritty, Windows Terminal and Console host on windows and also newest version of Alacritty on nixOS and Ubuntu. This issue can be triggered in one of two ways:

1. Set your prompt to something multiline via `$env.PROMPT_COMMAND = {|| "\n1\n2\n>" }` Fill entire screen with stuff (e.g. use `1..100`. Observer missing lines near edge of window.
2. Have a single line prompt. Fill like half or more of screen with stuff. Press f1 or ctrl+R (or open any other menu). Again observe missing lines near edge of window.

Some debugging shows that this (at least on platforms and terminals i tested) is not the same issue of missing history like in [589](https://invent.kde.org/utilities/konsole/-/merge_requests/589) mentioned earlier because even scroll of 1 deletes history. Thus any of  my attempts at using `ScrollUp(1)` failed. 

However I noticed that scrolling produced by printing "\n" on last line works just fine everywhere (thank god). Thus i suggest this simple fix [here](https://github.com/nushell/reedline/blob/2142c5c8e98aac8ce09d443b17a94f589b75afde/src/painting/painter.rs#L159) and [here](https://github.com/nushell/reedline/blob/2142c5c8e98aac8ce09d443b17a94f589b75afde/src/painting/painter.rs#L159) that sidesteps using scrolling entirely by replacing usage of `self.stdout.queue(ScrollUp(extra))?;` with:
```rust
self.stdout.queue(cursor::MoveTo(0, screen_height - 1))?;
for i in 0..extra {
    self.stdout.queue(Print(&coerce_crlf("\n")))?;
}
```

First the cursor is moved to the last line of window. This way if scroll is needed when cursor has not yet reached end of screen (e.g. when menu is rendered) it still works. Second scrolling is done via suitable number of line breaks (not entirely sure why, but doing it in one print does not work). Last the cursor must be moved in position where it was before, but in two places in reedline mentioned earlier this is not needed since cursor is moved immediately after anyways.

In my testing this fixes everything just fine. I can easily see how this is too hackish to merge for everyone. If not I can create a PR and do some more tests

**Comment by j4james** [comment]
> Some debugging shows that this (at least on platforms and terminals i tested) is not the same issue of missing history like in [589](https://invent.kde.org/utilities/konsole/-/merge_requests/589) mentioned earlier because even scroll of 1 deletes history.

@NotLebedev The issue was slightly different on Konsole, because a scroll of 1 did actually work (this was explained in the original [bug report](https://bugs.kde.org/show_bug.cgi?id=449233)). For most other terminals with this bug, the history is never retained, regardless of the scroll count, but other than that, it's essentially the same problem.

**Comment by fdncred** [comment]
@NotLebedev Thanks and good write-up! I think we're willing to take your idea for a spin and try it out if you're willing to submit a PR to reedline and nushell. It would also be great to back it up with some tests like you suggested.

**Comment by sholderbach** [comment]
This should be closed by nushell/reedline#601
